internal/http.go

```diff
@@ -0,0 +1,239 @@
+package internal
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/operate-first/curator-operator/pkg/signals"
+	"github.com/spf13/viper"
+
+	"github.com/jackc/pgx/v4"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+// Server defines the http server configuration
+type Server struct {
+	Addr        string
+	ErrorLog    *log.Logger
+	BaseContext func(net.Listener) context.Context
+}
+
+// HTTPError defines the http error codes
+type HTTPError struct {
+	code    int
+	message string
+}
+
+func InitializeRouter(ctx context.Context) {
+	var reportLog = ctrl.Log.WithName("report")
+	http.HandleFunc("/download", downloadReport)
+
+	// viper.SetDefault("HTTP.PORT", ":5000")
+	httpServerPort := viper.Get("HTTP.PORT").(string)
+
+	server := http.Server{
+		Addr:        httpServerPort,
+		BaseContext: func(_ net.Listener) context.Context { return ctx },
+	}
+	reportLog.Info("starting http server on port" + httpServerPort)
+
+	go startServer(&server, reportLog)
+
+	stopCh, closeCh := signals.CreateChannel()
+	defer closeCh()
+	log.Println("system call notified:", <-stopCh)
+
+	shutdownServer(ctx, &server, reportLog)
+}
+
+func startServer(server *http.Server, reportLog logr.Logger) {
+	if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
+		log.Fatal(err)
+	} else {
+		reportLog.Info("http server stopped gracefully")
+	}
+}
+
+func shutdownServer(ctx context.Context, server *http.Server, reportLog logr.Logger) {
+	ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+	defer cancel()
+
+	if err := server.Shutdown(ctx); err != nil {
+		reportLog.Error(err, "http server shutdown error:")
+	} else {
+		reportLog.Info("http server shutdowned")
+	}
+}
+
+func downloadReport(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	var reportLog = ctrl.Log.WithName("report")
+	HTTPError := validaeRequest(r, reportLog)
+	if HTTPError != nil {
+		http.Error(w, HTTPError.message, HTTPError.code)
+		return
```

Comment:
Why not return an error explicitly, and let any call site handle this edge case?